### PR TITLE
Support GitLab CI pipelines as source for the source-up-to-dateness metric

### LINKED ISSUE #3927: Allow for using a GitLab CI-pipeline as source for the 'source up-to-dateness' metric
GitLab already supports the source-up-to-dateness metric for files in GitLab. Use the same solution as for Azure DevOps: present both file path and pipeline parameters to the user and indicate in the tooltip that files get preference.

Note: Jenkins and Azure DevOps don't need to be changed. In case of Jenkins the source up-to-dateness of jobs can be measured (job artifacts have the same date, so no need to configure artifacts separately). In case of Azure DevOps Quality-time already supports both files and pipelines.

Reported by @Sebastiaan127001

### PR DESCRIPTION
**Pull request by fniessink**
Closes #3927.

### CODE REVIEW

File: components/collector/src/source_collectors/gitlab/source_up_to_dateness.py [hunk 1]
```diff
@@ -68,3 +72,35 @@ async def _parse_value(self, responses: SourceResponses) -> Value:
...
+class GitLabPipelineUpToDateness(TimePassedCollector, GitLabJobsBase):
+    """Collector class to measure the up-to-dateness of a job/pipeline."""
+
+    async def _landing_url(self, responses: SourceResponses) -> URL:
+        """Override to return a landing URL for the pipeline."""
+        if responses and (json := await responses[0].json()):
+            return URL(json[0]["pipeline"]["web_url"])
+        return await super()._landing_url(responses)
+
+    async def _parse_source_response_date_time(self, response: Response) -> datetime:
```
Comment (wkoot) [inline components/collector/src/source_collectors/gitlab/source_up_to_dateness.py:86]: Is it not easier to return a `datetime.date` object, as we only report on amount of days anyway?

File: components/collector/tests/source_collectors/gitlab/test_source_up_to_dateness.py [hunk 2]
```diff
@@ -46,7 +50,22 @@ async def test_source_up_to_dateness_folder(self):
...
+        """Test that the age of a pipeline can be measured."""
+        self.set_source_parameter("file_path", "")
+        build_date = datetime.fromisoformat(self.gitlab_jobs_json[0]["created_at"].strip("Z")).date()
+        expected_age = (date.today() - build_date).days
+        with self.patched_client_session_head():
+            response = await self.collect(get_request_json_return_value=self.gitlab_jobs_json)
+        self.assert_measurement(response, value=str(expected_age), landing_url="https://gitlab/project/-/pipelines/1")
+
+    async def test_file_landing_url_on_failure(self):
+        """Test that the landing url is the API url when GitLab cannot be reached."""
```
Comment (wkoot) [inline components/collector/tests/source_collectors/gitlab/test_source_up_to_dateness.py:63]: Where is the default value for `file_path` set?

**Review by wkoot — APPROVED** [review]

File: components/collector/src/source_collectors/gitlab/source_up_to_dateness.py (hunk 1 above)
Comment (fniessink) [reply components/collector/src/source_collectors/gitlab/source_up_to_dateness.py:86]: It is mostly use to report duration in days, that's correct, but returning dates here would amount to premature rounding, so not changing.

File: components/collector/tests/source_collectors/gitlab/test_source_up_to_dateness.py (hunk 2 above)
Comment (fniessink) [reply components/collector/tests/source_collectors/gitlab/test_source_up_to_dateness.py:63]: In the superclass of the test class.

**Pull request merged**